use 10.8 licensing features, implement unlimited guest users

wkloucek opened the pull request.

jvillafanez commented on lib/Command/CheckActiveUsers.php:
```diff
@@ -101,8 +104,19 @@ protected function execute(InputInterface $input, OutputInterface $output) {
+		if ($isQNAP) {
+			return $licenseManager->askLicenseFor('core', 'getUserAllowance');
+		}
+		$license = new QnapLicense('');
+		return $license->getUserAllowance();
```
This is questionable. Creating a fake license to return the user allowance of that license, when there is no valid license.... doesn't look good.

If there is no license, the default value for the user allowance shouldn't be gotten from the license itself but from a different place.
I think there are 2 main alternatives:
* Use a constant in the `QnapLicense` class. The difference is that you won't create a license. You could place the constant anywhere, although it makes more sense for the license to be there. Note that it must be known by anything planning to know what the current user allowance is (which might be a pain). It's an easy solution which provides a little benefit.
* Create a license adapter and use it across the app. The idea is to adjust the calls that the app needs to do and hide the complexity to the rest of the components. In particular, the adapter could have a `isQnapLicense` and `getUserAllowance` methods. Basic implementation could be:
  ```
  public function getUserAllowance() {
    if ($this->isQnapLicense()) {
      return $this->licenseManager->askLicenseFor(......);
    } else {
      return 0;  // could use a constant instead of a hardcoded value.
    }
  }
  ```
  The usage would be as simple as
  ```
  return $licenseAdapter->getUserAllowance();
  ```
  Note that there is no need to create a fake license anywhere.

wkloucek replied:
I went with the `QnapLicense::MIN_USER_ALLOWANCE` approach. This is also more clean in the way that the license parser doesn't hold this property anymore.

wkloucek commented on lib/QnapLicense.php:
```diff
@@ -2,23 +2,22 @@
 
 	public function getLicenseString(): string {
-		return 'qnap-license';
+		return ''; // license is stored by ownCloud not by QnapLicense
```
Ok , I reverted this change. I didn't know that some DBs have problems with empty strings

jvillafanez commented:
In general, I'd like to use and abuse the dependency injection because it make testing easier. It seems that we lack unit tests, so I'm not sure how strict we want to be with this. It will require a bunch of changes though.

Other than that, there are minor style changes which aren't important and can be overlooked.

In general, I think this is good enough to merge if we don't want to be strict with the code.

wkloucek commented:
> In general, I'd like to use and abuse the dependency injection because it make testing easier. It seems that we lack unit tests, so I'm not sure how strict we want to be with this. It will require a bunch of changes though.
> 
> Other than that, there are minor style changes which aren't important and can be overlooked.
> 
> In general, I think this is good enough to merge if we don't want to be strict with the code.

There is at least a test for the license parser part. But looks like it does not run in CI, I need to investigate in this. What other tests do you have in mind?

wkloucek commented:
@jvillafanez I activated unit tests in https://github.com/owncloud/qnap/pull/19. This produced some failing tests in this PR since the logic changed, but I already fixed them. I also added one test for the QnapLicense but to add more I would need some assistance (how to inject the timeFactory into the LicenseParser when creating a QnapLicense).

jvillafanez commented:
> I also added one test for the QnapLicense but to add more I would need some assistance (how to inject the timeFactory into the LicenseParser when creating a QnapLicense).

I can think 2 options:

Option1
Add a method `setLicenseParser(...)` that the test will use to inject the mock. The qnap code could be something like:
```
public function __construct(string $license) {
  $this->setLicenseParser(new LicenseParser(...), $license);
}
public function setLicenseParser($licenseParser, $licenseKey) {
  $this->licenseParser = $licenseParser;
  $this->licenseParser->loadLicensesText($licenseKey);
}
```
test code is expected to be something like:
```
protected function setUp(): void {
  $this->licenseParser = $this->createMock(LicenseParser::class);
  $this->qnapLicense = new QnapLicense('blblbl');
  $this->qnapLicense->setLicenseParser($this->licenseParser, 'fakeLicense');
}
```

Option2
Use reflection to inject the license parser mock into the private property. I'm not sure if there are easier alternatives such a assigning the mock directly (as I've read so far, it isn't possible).

In this case, there is no need to change the code, but using reflection isn't common.


There could be other options, although they require deeper changes.
I think that initially, the license was thought as a data holder, so it shouldn't have any dependency. This is why injecting a dependency is difficult or it wasn't planned.
In this context, you could have a license loader to "pre-parse" the license, and then the license itself which could do something easier which doesn't require external code. As said, this requires quite a bunch of changes.


I guess option2 is the best one in this case.

jvillafanez commented on tests/unit/QnapLicenseTest.php:
```diff
@@ -0,0 +1,180 @@
+
+	protected function setUp(): void {
+		parent::setUp();
+		$timeFactory = $this->createMock(ITimeFactory::class);
```
you should mock direct dependencies of the class being tested. In this case, it's the `LicenseParser` class the one that should be mocked.
We want to test just the code in the `QnapLicense` class, not anything else. For this case, it's pretty easy since you just need to check that the `QnapLicense` class returns the same things the `LicenseParser` does (it's a direct call for the most of the methods).

Most of the stuff here seems to try to test the `LicenseParser`, but those tests should be in the `LicenseParserTest` class

wkloucek replied:
Thanks, this made it a lot easier

jvillafanez commented:
The command and the panel should also be tested. Furthermore, use and abuse dependency injection for both of them because it make testing easier.

jvillafanez commented on tests/unit/QnapLicenseTest.php:
```diff
@@ -0,0 +1,62 @@
+		$this->parser->method('isValid')->willReturn($givenValid);
+		$this->parser->method('getUserAllowance')->willReturn($givenUsers);
+
+		$p = $ref->getProperty('licenseParser');
+		$p->setAccessible(true);
+		$p->setValue($license, $this->parser);
```
This is part of the test setup, so it should be in the `setUp` method.

jvillafanez commented on tests/unit/QnapLicenseTest.php:
```diff
@@ -0,0 +1,62 @@
+	 * @param int $expectedExpiration
+	 * @param string $licenseStr
+	 */
+	public function test(bool $givenValid, int $givenUsers, bool $expectedValid, int $expectedUsers, int $expectedType): void {
```
There should be at least one test per public method.
I'd expect a `testGetLicenseString`, `testIsValid`, `testGetExpirationTime`, etc. Data providers can help to cover all the cases, but you could include more tests if it becomes complex.

jvillafanez commented on tests/unit/QnapLicenseTest.php:
```diff
@@ -0,0 +1,62 @@
+
+use OCA\QNAP\QnapLicense;
+use OCA\QNAP\LicenseParser;
+use OCP\AppFramework\Utility\ITimeFactory;
```
This isn't needed any longer

jvillafanez commented on tests/unit/QnapLicenseTest.php:
```diff
@@ -0,0 +1,102 @@
+		$this->parser->method('getUserAllowance')->willReturn(0);
+		self::assertEquals(5, $this->license->getUserAllowance());
+
+		$this->parser->method('getUserAllowance')->willReturn(5);
```
these 3 tests fit better using a data provider, like `getUserAllowance($parsedAllowance, $expectedAllowance)`

jvillafanez commented on tests/unit/QnapLicenseTest.php:
```diff
@@ -0,0 +1,102 @@
+	 * @param int $expectedExpiration
+	 * @param string $licenseStr
+	 */
+	public function testLicense(bool $givenValid, int $givenUsers, bool $expectedValid, int $expectedUsers, int $expectedType): void {
```
this test shouldn't be needed because all the test cases should be already covered with the rest of the tests. Up to you to decide what to do.

jvillafanez commented:
the command and the panel should also have their own unit tests. You should adjust the code first in order to inject the dependencies though

wkloucek commented on tests/unit/AdminPanelTest.php:
```diff
@@ -0,0 +1,60 @@
+		$this->createUser('user4');
+		$this->createUser('user5');
+
+		// TODO: create guest users too
```
@jvillafanez is there an easy way to create guest users or do I need to mock the user manager?

jvillafanez replied:
All the dependencies of the class should be mocked. The only classes that might not be mocked are models or data classes, which could be considered as a custom data type (usually having just get / set methods for the data).

For this case, yes, you should mock the users you need and the mock the user manager to return (or use) the list of mocked users.

jvillafanez replied:
Note that you should be testing only public functions. Testing private functions shouldn't be needed unless the code is too complex, in which case, we should consider to move that complex code to another class.
Of course, all the code (including private functions) should be covered with unit tests.

jvillafanez commented on lib/AdminPanel.php:
```diff
@@ -2,24 +2,52 @@
+	 */
+	private $userManager;
+
+	public function __construct() {
```
```
public function __construct(ILicenseManager $licenseManager, IUserManager $userManager) {
  $this->licenseManager = $licenseManager;
  $this->userManager = $userManager;
}
```
The injection of the actual objects should be done automatically by ownCloud.

jvillafanez commented on tests/unit/AdminPanelTest.php:
```diff
@@ -0,0 +1,60 @@
+	}
+
+	public function testGetFiveUserCount(): void {
+		$class = new \ReflectionClass($this->panel);
```
This was an easy workaround for the QnapLicense test. It shouldn't be used normally.
Dependencies should be mocked and injected in the `setUp()` method of the tests. The code should be prepared for this.

jvillafanez commented on lib/AdminPanel.php:
```diff
@@ -29,12 +57,27 @@ public function getSectionID() {
+		return $numberOfActiveUsers - $this->getGuestUserCount();
+	}
+
+	private function getGuestUserCount(): int {
```
It seems I overlooked this....

Please, consider to add the function code inside the `getUserCount` function. You're traversing the full list of users twice, which could be expensive.
Taking into account the `getUserCount` is private, you can return an array `['normal users' => 35, 'guest users' => 1]`. You can handle the counters in one callback function and traverse the list only once.

jvillafanez commented on tests/unit/AdminPanelTest.php:
```diff
@@ -0,0 +1,145 @@
+	}
+
+	public function testGetPanel(): void {
+		$tmpl = new \ReflectionClass(Template::class);
```
you could do something like
```
$html = $this->panel->getPanel()->fetchPage();
```
and then assert that the html contains the strings we want. This way you don't need to fiddle with the internals of the class.
There is no need to be strict and match the exact html for the panel. Just checking for the interesting parts is enough to keep some flexibility in case the template changes.

jvillafanez commented on lib/AdminPanel.php:
```diff
@@ -2,39 +2,78 @@
+				if ($this->userTypeHelper->isGuestUser($user->getUID()) === true) {
+					$numberOfActiveGuestUsers++;
+				} else {
+					$numberOfActiveUsers++;
```
we might need a comment somewhere to ensure we're counting non-guest users. This has been changed here and we don't want more changes after the code is stable.

wkloucek replied:
is a bug in the current implementation: https://github.com/owncloud/qnap/issues/17

jvillafanez commented on tests/unit/CheckActiveUsersTest.php:
```diff
@@ -0,0 +1,279 @@
+		$input = $this->createMock(InputInterface::class);
+		$output = $this->createMock(OutputInterface::class);
+
+		$cmd = new \ReflectionClass(CheckActiveUsers::class);
```
Better use symfony's command tester. You can check an example in https://github.com/owncloud/core/blob/eebd754771a99bc44820eb67a6d57cea238e6b86/apps/files_trashbin/tests/Command/ExpireTrashTest.php

jvillafanez commented on tests/unit/CheckActiveUsersTest.php:
```diff
@@ -0,0 +1,279 @@
+		self::assertEquals(1, \count($this->mails));
+	}
+
+	public function testExecuteWithinUserAllowance(): void {
```
You could use a data provider containing the current users and the expected data, where the expected data contains the active users, guests users, etc.

jvillafanez commented on tests/unit/CheckActiveUsersTest.php:
```diff
@@ -0,0 +1,279 @@
+				return $this->createMock(INotification::class);
+			})
+		);
+		$this->notificationManager->method("notify")->will(
```
This is too big for a setup method. Normally, the setup method just creates the mocks and the test object, nothing else. Further initialization such as mock configuration is done inside the test because it depends on each test.

You could create an "init(....)" method or similar to "finish the setup", but it's important that this is called inside the test method because, as said, the configuration might be different for each test.

In addition, note that you can set expectations to check how many times a mocked method is called:
```
$this->notificationManager->expects($this->once())->method("notify");
$this->notificationManager->expects($this->never())->method("notify");
$this->notificationManager->expects($this->exactly(3))->method("notify");
```
This seems a cleaner approach.

jvillafanez commented:
I'm still not fully convinced with the tests for the command, but I don't have easier and better options at the moment.
Maybe, taking advantage of the command tester, we could check the output instead of doing operations in the test. Basically, we'd be more interested in checking that there is a message saying that there are 60 normal users, instead of counting those users ourselves. Of course, we're assuming that the command won't lie.
The problem I see is that the test leaks the user information. The `$this->enabledUsers` variable (and others) isn't cleaned up after the test, which could cause a lot of confusion if we need to include more tests there. You could be using data from a different test.

Other than that, I don't want to block the code too much, so if you're confident with the result, I'll approve the changes.

wkloucek commented:
> I'm still not fully convinced with the tests for the command, but I don't have easier and better options at the moment.
> Maybe, taking advantage of the command tester, we could check the output instead of doing operations in the test. Basically, we'd be more interested in checking that there is a message saying that there are 60 normal users, instead of counting those users ourselves. Of course, we're assuming that the command won't lie.

For our use case it is essential, that the excess users are properly disabled (and therefore that the counting is right). That's why I prefer checking this directly and not just the command output.

> The problem I see is that the test leaks the user information. The `$this->enabledUsers` variable (and others) isn't cleaned up after the test, which could cause a lot of confusion if we need to include more tests there. You could be using data from a different test.

That's a good point. For that I added a cleanup step (`tearDown()` https://phpunit.readthedocs.io/en/9.5/fixtures.html?highlight=teardown#fixtures).

> Other than that, I don't want to block the code too much, so if you're confident with the result, I'll approve the changes.

Test coverage wise we are now way better than before and I really appreciate your help on that :-) I think we can go on confidently
